### PR TITLE
[WIP] Multicore supports

### PR DESCRIPTION
**Pull request by zhangxiaomei**
1. Add multicore supports in SiteDirector, and MultiProcessorSiteDirector can be removed
    Sites can use three parameters to define multi-core features: NumberOfProcessors, WholeNode, RequiredTag. Without any these features, sites are used as singlecore. 
    These features can be used like the following:
        CEs
        {
          slurm.ihep.ac.cn
          {
           Queues
            {
              sdirac
              { 
                ......
                RequiredTag = 2Processors
                ......
              }
              smdirac
              { 
                ......
                WholeNode = true
                RequiredTag = 2Processors
              }
            NumberOfProcessors = 4
          }      
2. Realize multicore supports to SLURM and Condor
    Condor still need adding "WholeNode" Support

### CODE REVIEW

File: WorkloadManagementSystem/Agent/SiteDirector.py [hunk 1]
```diff
@@ -626,7 +685,16 @@ def submitJobs( self ):
           return result
 
         executable, pilotSubmissionChunk = result['Value']
-        result = ce.submitJob( executable, '', pilotSubmissionChunk )
+       
+        #check the parameter
+        processors = self.pilotProcessors
+        if self.wholeNode == 'true':
+          processors = -1
+        if processors != 1:
```
Comment (pigay) [inline WorkloadManagementSystem/Agent/SiteDirector.py]: Calling `ce.submitJob( executable, '', pilotSubmissionChunk, processors = processors )` directly without the test should be better?

File: Resources/Computing/BatchSystems/Condor.py [hunk 2]
```diff
@@ -81,22 +81,23 @@ def submitJob( self, **kwargs ):
...
     executable = kwargs['Executable']
     submitOptions = kwargs['SubmitOptions']
+    pn = kwargs['Mprocess']
 
     jdlFile = tempfile.NamedTemporaryFile( dir=outputDir, suffix=".jdl" )
     jdlFile.write("""
     Executable = %s
     Universe = vanilla
-    Requirements   = OpSys == "LINUX"
+    Requirements = Target.OpSys == "LINUX"
```
Comment (fstagni) [inline Resources/Computing/BatchSystems/Condor.py]: What is the rationale behind this?

File: WorkloadManagementSystem/PilotAgent/pilotCommands.py [hunk 3]
```diff
@@ -484,7 +484,7 @@ def execute( self ):
         self.cfg.append( '-ddd' )
 
       if numberOfProcessor:
-        self.cfg.append( '-o "/Resources/Computing/CEDefaults/NumberOfProcessors=%d"' % numberOfProcessor )
+        self.cfg.append( '-o "/Resources/Computing/CEDefaults/SiteProcessors=%d"' % numberOfProcessor )
```
Comment (fstagni) [inline WorkloadManagementSystem/PilotAgent/pilotCommands.py]: Why this change?

File: WorkloadManagementSystem/Agent/SiteDirector.py (hunk 1 above)
Comment (zhangxiaomei) [inline WorkloadManagementSystem/Agent/SiteDirector.py]: We still need time to cover all the CE types, so keep the old function to support the single-core-only CEs we haven't covered. It will be removed after all the CE types can support multi-core in DIRAC.

File: Resources/Computing/BatchSystems/Condor.py [hunk 4]
```diff
@@ -67,7 +67,7 @@ def submitJob( self, **kwargs ):
 
     resultDict = {}
 
-    MANDATORY_PARAMETERS = [ 'Executable', 'OutputDir', 'SubmitOptions' ]
+    MANDATORY_PARAMETERS = [ 'Executable', 'OutputDir', 'SubmitOptions', 'Mprocess' ]
```
Comment (zhangxiaomei) [inline Resources/Computing/BatchSystems/Condor.py]: yes, you are right.  We will change it, not putting it as mandatory.

File: Resources/Computing/BatchSystems/Condor.py (hunk 2 above)
Comment (zhangxiaomei) [inline Resources/Computing/BatchSystems/Condor.py]: It is suggested by our condor admin that it is a standard way to describe "Requirements" with "Target".

File: WorkloadManagementSystem/PilotAgent/pilotCommands.py (hunk 3 above)
Comment (zhangxiaomei) [inline WorkloadManagementSystem/PilotAgent/pilotCommands.py]: In multi-core case, "/Resources/Computing/CEDefaults/NumberOfProcessors=%d" has been used in PoolComputingElement to get the total processors which the current multi-core pilot holds and count current available cores of the pilot.  It is conflicted with "NumberOfProcessors" in "PilotCommand.py", which is to describe the total cores which the WN holds and not used in other places now.  Therefore, we choose to change the parameter in PilotCommand.py.

**Comment by fstagni** [comment]
This branch has conflicts, which should be resolved. 

At the same time there are other comments to this PR that has not yet been taken into account. I am positive about merging this once current issues have been sorted out.

**Comment by zhangxiaomei** [comment]
Part of codes have already been merged to Andrei's PR. eg. the changes in SiteDirector.py.  The rest of part I prefer to open another PR to continue. What do you suggest?

**Comment by fstagni** [comment]
I think it's a good idea to open another PR, based on what's the content of the integration branch right now. Please go ahead, and close this PR once the new one is ready.

**Comment by zhangxiaomei** [comment]
This PR has been continued in PR#3441 and PR#3503
